### PR TITLE
[tmva] RBatchGenerator: fix chunking, add tests, add drop_remainder parameter

### PR DESCRIPTION
**Pull request by kristupaspranc**
Change TFDatasets to TFGenerators.
Add iteration with next() option.
Return target and weights data in columns.
Minor fixes.

### CODE REVIEW

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def get_template(
             name_str = str(name)
             self.given_columns.append(name_str)
             column_type = template_dict[str(x_rdf.GetColumnType(name_str))]
-            template_string += column_type + ","
+            template_string = f"{template_string}{column_type},"
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:82]: I wonder what is the reasoning behind this change, as it changes the logic in place. The variable `template_string` was previously declared out of the loop and accumulated at each iteration. Now it's just defined every time anew inside the loop, making its definition at line 70 superfluous. Can you elaborate more here?

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 2]
```diff
@@ -158,6 +159,9 @@ def __init__(
             shuffle (bool):
                 Batches consist of random events and are shuffled every epoch.
                 Defaults to True.
+            drop_remainder (bool):
+                Drop the remainder of data that is too small to compose full batch.
+                Defaults to True.
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:164]: A bit more documentation here with a small example would be nice

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 3]
```diff
@@ -306,18 +311,21 @@ def ConvertBatchToNumpy(self, batch: "RTensor") -> np.ndarray:
...
 
         data = batch.GetData()
-        data.reshape((self.batch_size * self.num_columns,))
 
-        return_data = np.array(data).reshape(self.batch_size, self.num_columns)
+        if tuple(batch.GetShape()) != (self.batch_size,self.num_columns):
+            batch_size, num_columns = tuple(batch.GetShape())
+        else:
+            batch_size = self.batch_size
+            num_columns = self.num_columns
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:319]: Why is this done unconditionally? What is a situation where the first `if` condition might be true? e.g. when is the shape of the batch different from the stored attributes? This practically means that what the user asked initially is not what they will get

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 4]
```diff
@@ -306,18 +311,21 @@ def ConvertBatchToNumpy(self, batch: "RTensor") -> np.ndarray:
...
+
+        data.reshape((batch_size * num_columns,))
+
+        return_data = np.array(data).reshape(batch_size, num_columns)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:323]: Are we sure this does not copy the array?

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 5]
```diff
@@ -326,15 +334,10 @@ def ConvertBatchToNumpy(self, batch: "RTensor") -> np.ndarray:
...
 
                 weights_data = return_data[:, self.weights_index]
-                return_data = np.column_stack(
-                    (
-                        return_data[:, : self.weights_index],
-                        return_data[:, self.weights_index + 1 :],
-                    )
-                )
-                return return_data, target_data, weights_data
+                return_data = np.column_stack((return_data[:, : self.weights_index], return_data[:, self.weights_index + 1 :]))
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:337]: This means that the `return_data` is computed twice in case the `if self.weights_given` is true. The logic can be improved

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 6]
```diff
@@ -350,56 +353,78 @@ def ConvertBatchToPyTorch(self, batch: Any) -> torch.Tensor:
...
-        batch = self.ConvertBatchToNumpy(batch)
+        data = batch.GetData()
 
-        # TODO: improve this by returning tensorflow tensors
-        return batch
+        if tuple(batch.GetShape()) != (self.batch_size,self.num_columns):
+            batch_size, num_columns = tuple(batch.GetShape())
+        else:
+            batch_size = self.batch_size
+            num_columns = self.num_columns
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:404]: Similarly here, I fear that this may result in unexpected behaviour for the user in case they don't see the correct batch size returned to them

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 7]
```diff
@@ -350,56 +353,78 @@ def ConvertBatchToPyTorch(self, batch: Any) -> torch.Tensor:
...
+            if self.weights_given:
+                if self.target_index < self.weights_index:
+                    self.weights_index -= 1
+
+                weights_data = return_data[:, self.weights_index]
+                return_data = tf.concat([return_data[:, : self.weights_index], return_data[:, self.weights_index + 1 :]], axis=1)
+
+                return return_data, tf.expand_dims(target_data,axis=1), tf.expand_dims(weights_data,axis=1)
+
+            return return_data, tf.expand_dims(target_data,axis=1)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:425]: This logic is replicated almost exactly the same across three functions. It could probably be factored out somehow. Possibly for another PR though

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 8]
```diff
@@ -491,6 +516,9 @@ def __iter__(self):
         return self
 
     def __next__(self):
+        if not hasattr(self, '_callable'):
+            self._callable = self.__call__()
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:520]: Is there any situation where `self._callable` is something other than `self.__call__()`? Otherwise, this condition can just be removed and just use `self.__call__()` directly (aka `self()`)

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py [hunk 9]
```diff
@@ -677,6 +714,7 @@ def CreateTFDatasets(
     validation_split: float = 0.0,
     max_chunks: int = 0,
     shuffle: bool = True,
+    drop_remainder = True,
 ) -> Tuple[tf.data.Dataset, tf.data.Dataset]:
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:718]: This is now false

File: bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+            collected_y_train.append(y.tolist())
+        
+        for x, y in gen_validation:
+            collected_x_val.append(x.tolist())
+            collected_y_val.append(y.tolist())
+
+        flat_x_train = [x for xl in collected_x_train for xs in xl for x in xs]
+        flat_x_val = [x for xl in collected_x_val for xs in xl for x in xs]
+        flat_y_train = [y for yl in collected_y_train for ys in yl for y in ys]
+        flat_y_val = [y for yl in collected_y_val for ys in yl for y in ys]
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py:69]: As we discussed offline, these tests must also check the shape of every batch to be correct

File: bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        validation_split=0.3,
+        shuffle=False,
+        drop_remainder=False
+        )
+
+        for x, y in gen_train:
+            pass
+
+        for x, y in gen_validation:
+            pass
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py:190]: This test needs explanation about why you have an empty for loop. If you're just checking against some major failure (e.g. a segfault), probably you don't need a for loop to do that

File: bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    # def create_20_entries_file(self):
+    #     df1 = ROOT.RDataFrame(20)\
+    #         .Define("b1", "(int) rdfentry_ + 10")\
+    #         .Define("b2", "(double) rdfentry_ * rdfentry_")\
+    #         .Snapshot(self.tree_name, self.file_name2)
+        
+        #print(df1.Describe())
+        #print(df2.Describe())
+
+        #df2.Display("",).Print()
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py:27]: Leftover comments should go away

File: bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+
+
+    def teardown_file(self, file):
+        os.remove(file)
```
Comment (vepadulano) [inline bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py:31]: This function should not be there. setup and teardown routines for Python unittests are handled via `setup*` and `tearDown*` functions. See the documentation at https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUpClass

File: tutorials/tmva/RBatchGenerator_NumPy.py [hunk 14]
```diff
@@ -27,10 +27,22 @@
...
-for i, b in enumerate(ds_train):
-    print(f"Training batch {i} => {b.shape}")
+i = 1
+while True:
+    try:
+        b = next(ds_train)
+        print(f"Training batch {i} => {b.shape}")
+        i+=1
+    except StopIteration:
+        break
```
Comment (vepadulano) [inline tutorials/tmva/RBatchGenerator_NumPy.py:37]: This is really ugly to show in a tutorial, I would just revert this change

**Review by vepadulano — CHANGES_REQUESTED** [review]
Thank you for the nice progress! I have just reviewed the Python part of the PR, but let's start with the following comments

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py (hunk 1 above)
Comment (kristupaspranc) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:82]: It works as discussed offline

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py (hunk 2 above)
Comment (kristupaspranc) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:164]: Noted and done for the next PR

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py (hunk 3 above)
Comment (kristupaspranc) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:319]: Condition check will be removed

File: bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py (hunk 9 above)
Comment (kristupaspranc) [reply bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_batchgenerator.py:718]: I am not sure what is meant by that, thought we decided on having the parameter set to True as it is safer in case user misses the feature

File: tutorials/tmva/RBatchGenerator_NumPy.py (hunk 14 above)
Comment (kristupaspranc) [reply tutorials/tmva/RBatchGenerator_NumPy.py:37]: Reverted in the next PR

File: bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py (hunk 13 above)
Comment (kristupaspranc) [reply bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py:31]: Will improve, thanks!

File: bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py (hunk 12 above)
Comment (kristupaspranc) [reply bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py:27]: Was prepared for the next tests, noted

File: bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py (hunk 10 above)
Comment (kristupaspranc) [reply bindings/pyroot/pythonizations/test/rbatchgenerator_completeness.py:69]: Already done for the next PR
